File: apps/api/src/caseops_api/schemas/ip_patent_prosecution.py
"""Source-pinned patent work product and prosecution, not legal automation."""

from datetime import date
from typing import Literal, Self
from uuid import UUID

from pydantic import AwareDatetime, Field, model_validator

from caseops_api.schemas.ip_patents import (
    NonBlankText,
    PatentContract,
    PatentDocumentSource,
    PatentProsecutionPhase,
    Sha256,
)

PatentDocumentKind = Literal[
    "claims",
    "specification",
    "drawings",
    "abstract",
    "sequence_listing",
    "translation",
    "amendment",
    "response",
    "filing_package",
]
PatentEventKind = Literal[
    "filing_preparation",
    "filing",
    "publication",
    "examination_request",
    "office_action",
    "response",
    "hearing",
    "amendment",
    "grant",
    "restoration",
]


class PatentWorkPreconditions(PatentContract):
    expected_version: int = Field(strict=True, ge=1)
    expected_lifecycle_version: int = Field(strict=True, ge=0)
    expected_work_sequence: int = Field(strict=True, ge=0)
    reason: NonBlankText = Field(min_length=5, max_length=1000)


class PatentManifestItem(PatentContract):
    document_kind: PatentDocumentKind
    source: PatentDocumentSource


class PatentEvidenceCreateRequest(PatentWorkPreconditions):
    title: NonBlankText = Field(min_length=1, max_length=255)
    document_kind: PatentDocumentKind
    predecessor_id: UUID | None = None
    source: PatentDocumentSource
    documents: tuple[PatentManifestItem, ...] = Field(min_length=1, max_length=20)

    @model_validator(mode="after")
    def validate_manifest(self) -> Self:
        versions = [item.source.document_version_id for item in self.documents]
        if len(set(versions)) != len(versions):
            raise ValueError("A manifest cannot repeat a document version.")
        if self.document_kind != "filing_package" and (
            len(self.documents) != 1 or self.documents[0].document_kind != self.document_kind
        ):
            raise ValueError("A document edition pins one document of its declared kind.")
        if any(item.document_kind == "filing_package" for item in self.documents):
            raise ValueError("A filing manifest contains documents, not nested packages.")
        return self


class PatentEvidenceRecord(PatentContract):
    id: UUID
    application_id: UUID
    root_id: UUID
    predecessor_id: UUID | None
    sequence: int
    edition: int
    is_current: bool
    anchor_version: int
    lifecycle_version: int
    title: str
    document_kind: PatentDocumentKind
    source: PatentDocumentSource
    documents: tuple[PatentManifestItem, ...]
    manifest_sha256: Sha256
    reason: str
    created_at: AwareDatetime


class PatentProsecutionCreateRequest(PatentWorkPreconditions):
    event_kind: PatentEventKind
    received_on: date
    effective_on: date
    source: PatentDocumentSource
    evidence_id: UUID | None = None
    expected_phase: PatentProsecutionPhase
    exceptional_transition_reason: NonBlankText | None = Field(default=None, max_length=1000)
    preview_sha256: Sha256 | None = None
    acknowledged_exception_codes: tuple[
        Literal[
            "backdated_recalculation_review_required", "exceptional_transition_review_required"
        ],
        ...,
    ] = Field(default=(), max_length=2)


class PatentProsecutionPreview(PatentContract):
    application_id: UUID
    work_sequence: int
    current_phase: PatentProsecutionPhase
    proposed_phase: PatentProsecutionPhase
    backdated: bool
    affected_deadline_ids: tuple[UUID, ...]
    required_acknowledgements: tuple[str, ...]
    preview_sha256: Sha256
    changes_deadlines: Literal[False] = False
    authoritative_calculation_available: Literal[False] = False


class PatentProsecutionRecord(PatentContract):
    id: UUID
    application_id: UUID
    sequence: int
    anchor_version: int
    lifecycle_version: int
    event_kind: PatentEventKind
    received_on: date
    effective_on: date
    source: PatentDocumentSource
    evidence_id: UUID | None
    before_phase: PatentProsecutionPhase
    after_phase: PatentProsecutionPhase
    reason: str
    exceptional_transition_reason: str | None
    impact: PatentProsecutionPreview
    created_at: AwareDatetime


class PatentEvidencePage(PatentContract):
    application_id: UUID
    work_sequence: int
    records: list[PatentEvidenceRecord]
    next_cursor: int | None


class PatentProsecutionPage(PatentContract):
    application_id: UUID
    work_sequence: int
    records: list[PatentProsecutionRecord]
    next_cursor: int | None

File: apps/web/lib/api/ip-patent-prosecution.ts
import { z } from "zod";

import { apiRequest } from "./client";
import { patentApplicationSchema } from "./ip-patents";

const hash = z.string().regex(/^[a-f0-9]{64}$/);
const source = z.object({
  kind: z.literal("document_version"), document_id: z.string().uuid(),
  document_version_id: z.string().uuid(), content_sha256: hash,
}).strict();
export const patentDocumentKinds = ["claims", "specification", "drawings", "abstract", "sequence_listing", "translation", "amendment", "response", "filing_package"] as const;
export const patentEventKinds = ["filing_preparation", "filing", "publication", "examination_request", "office_action", "response", "hearing", "amendment", "grant", "restoration"] as const;
const manifestItem = z.object({ document_kind: z.enum(patentDocumentKinds), source }).strict();
const record = {
  id: z.string().uuid(), application_id: z.string().uuid(), sequence: z.number().int().positive(),
  anchor_version: z.number().int().positive(), lifecycle_version: z.number().int().nonnegative(),
  reason: z.string(), created_at: z.string().datetime({ offset: true }), source,
};
export const patentEvidenceSchema = z.object({
  ...record, root_id: z.string().uuid(), predecessor_id: z.string().uuid().nullable(),
  edition: z.number().int().positive(), is_current: z.boolean(), title: z.string(), document_kind: z.enum(patentDocumentKinds),
  documents: z.array(manifestItem).min(1).max(20), manifest_sha256: hash,
}).strict();
export const patentProsecutionPreviewSchema = z.object({
  application_id: z.string().uuid(), work_sequence: z.number().int().nonnegative(),
  current_phase: patentApplicationSchema.shape.prosecution_phase,
  proposed_phase: patentApplicationSchema.shape.prosecution_phase,
  backdated: z.boolean(), affected_deadline_ids: z.array(z.string().uuid()).max(100),
  required_acknowledgements: z.array(z.enum(["backdated_recalculation_review_required", "exceptional_transition_review_required"])).max(2),
  preview_sha256: hash, changes_deadlines: z.literal(false), authoritative_calculation_available: z.literal(false),
}).strict();
export const patentProsecutionSchema = z.object({
  ...record, event_kind: z.enum(patentEventKinds), received_on: z.string().date(), effective_on: z.string().date(),
  evidence_id: z.string().uuid().nullable(), before_phase: patentApplicationSchema.shape.prosecution_phase,
  after_phase: patentApplicationSchema.shape.prosecution_phase, exceptional_transition_reason: z.string().nullable(),
  impact: patentProsecutionPreviewSchema,
}).strict();
const evidencePage = z.object({ application_id: z.string().uuid(), work_sequence: z.number().int().nonnegative(),
  records: z.array(patentEvidenceSchema).max(100), next_cursor: z.number().int().positive().nullable() }).strict();
const eventPage = z.object({ application_id: z.string().uuid(), work_sequence: z.number().int().nonnegative(),
  records: z.array(patentProsecutionSchema).max(100), next_cursor: z.number().int().positive().nullable() }).strict();

export type PatentEvidence = z.infer<typeof patentEvidenceSchema>;
export type PatentProsecution = z.infer<typeof patentProsecutionSchema>;
export type PatentProsecutionPreview = z.infer<typeof patentProsecutionPreviewSchema>;
type Preconditions = { expected_version: number; expected_lifecycle_version: number; expected_work_sequence: number; reason: string };
export type PatentEvidenceInput = Preconditions & {
  title: string; document_kind: typeof patentDocumentKinds[number]; predecessor_id?: string | null;
  source: z.infer<typeof source>; documents: z.infer<typeof manifestItem>[];
};
export type PatentEventInput = Preconditions & {
  event_kind: typeof patentEventKinds[number]; received_on: string; effective_on: string;
  source: z.infer<typeof source>; evidence_id: string | null;
  expected_phase: z.infer<typeof patentApplicationSchema>["prosecution_phase"];
  exceptional_transition_reason: string | null; preview_sha256?: string;
  acknowledged_exception_codes?: PatentProsecutionPreview["required_acknowledgements"];
};
const path = (applicationId: string) => `/api/ip/patents/applications/${encodeURIComponent(applicationId)}`;
const pageQuery = (cursor?: number, snapshot?: number) => {
  const query = new URLSearchParams({ limit: "25" });
  if (cursor !== undefined) query.set("cursor", String(cursor));
  if (snapshot !== undefined) query.set("snapshot_sequence", String(snapshot));
  return query;
};
export async function fetchPatentEvidence(id: string, cursor?: number, snapshot?: number, signal?: AbortSignal) {
  return evidencePage.parse(await apiRequest<unknown>(`${path(id)}/evidence?${pageQuery(cursor, snapshot)}`, { signal }));
}
export async function fetchPatentProsecution(id: string, cursor?: number, snapshot?: number, signal?: AbortSignal) {
  return eventPage.parse(await apiRequest<unknown>(`${path(id)}/prosecution?${pageQuery(cursor, snapshot)}`, { signal }));
}
export async function createPatentEvidence(id: string, body: PatentEvidenceInput, key: string) {
  return patentEvidenceSchema.parse(await apiRequest<unknown>(`${path(id)}/evidence`, { method: "POST", body, headers: { "Idempotency-Key": key } }));
}
export async function previewPatentProsecution(id: string, body: PatentEventInput) {
  return patentProsecutionPreviewSchema.parse(await apiRequest<unknown>(`${path(id)}/prosecution/preview`, { method: "POST", body }));
}
export async function createPatentProsecution(id: string, body: PatentEventInput, key: string) {
  return patentProsecutionSchema.parse(await apiRequest<unknown>(`${path(id)}/prosecution`, { method: "POST", body, headers: { "Idempotency-Key": key } }));
}
